test/migration/migration_test.go

```diff
@@ -0,0 +1,219 @@
+		log.Printf("failed to setup framework: %v\n", err)
+		t.Fatal(err)
+	}
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+	ns2 := ctx.CreateNamespace(t, framework.KubeClient)
+	ctx.SetupPrometheusRBAC(t, ns2, framework.KubeClient)
+
+	// Launch the objects.
+	name := "test"
+	group := "servicediscovery-test"
+
+	prometheusTPR := framework.MakeBasicPrometheus(ns2, name, name, 1)
+	prometheusTPR.Namespace = ns2
+	if err := framework.CreatePrometheusAndWaitUntilReady(ns2, prometheusTPR); err != nil {
+		t.Fatal(err)
+	}
+
+	s := framework.MakeBasicServiceMonitor(group)
+	if _, err := framework.MonClient.ServiceMonitors(ns2).Create(s); err != nil {
+		t.Fatal("Creating ServiceMonitor failed: ", err)
+	}
+
+	if err := framework.CreateAlertmanagerAndWaitUntilReady(ns2, framework.MakeBasicAlertmanager(name, 3)); err != nil {
+		t.Fatal(err)
+	}
+
+	obj, err := monClient.Prometheuses(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	proms := obj.(*v1alpha1.PrometheusList)
+
+	// Get the objects.
+	obj, err = monClient.Alertmanagers(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	ams := obj.(*v1alpha1.AlertmanagerList)
+
+	obj, err = monClient.ServiceMonitors(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	sms := obj.(*v1alpha1.ServiceMonitorList)
+
+	// Delete and launch new operator.
+	deletePolicy := metav1.DeletePropagationForeground
+	if err := framework.
+		KubeClient.
+		Extensions().
+		Deployments(framework.Namespace.Name).
+		Delete("prometheus-operator", &metav1.DeleteOptions{
+			PropagationPolicy: &deletePolicy,
+		}); err != nil {
+		t.Fatal(err)
+	}
+
+	wait.Poll(time.Second, 3*time.Minute, func() (bool, error) {
+		_, err := framework.KubeClient.
+			AppsV1beta1().Deployments(framework.Namespace.Name).
+			Get("prometheus-operator", metav1.GetOptions{})
+
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+
+			return false, err
+		}
+
+		return false, nil
+	})
+
+	if err := framework.Setup(*opImage); err != nil {
+		t.Fatal(err)
+	}
+
+	// Check if TPRs are deleted.
+	tprList, err := tprClient.List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(tprList.Items) != 0 {
+		t.Fatalf("expected 0 TPRs got %d", len(tprList.Items))
+	}
+
+	// Check if CRDs are created.
+	crdList, err := crdClient.List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(crdList.Items) != 3 {
+		t.Fatalf("expected 3 TPRs got %d", len(crdList.Items))
+	}
+
+	// Compare old and new objects.
+	obj, err = monClient.Prometheuses(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	promsNew := obj.(*v1alpha1.PrometheusList)
+	if len(promsNew.Items) != len(proms.Items) {
+		t.Fatalf("expected %d proms, got %d", len(proms.Items), len(promsNew.Items))
+	}
+
+	for i, prom := range proms.Items {
+		if !reflect.DeepEqual(prom.Spec, promsNew.Items[i].Spec) {
+			t.Fatalf("The prometheus object changed %d", i)
```

Comment:
I think the comments are lacking. Will send in a PR with more comments and maybe breaking it into smaller ones today.